06_gpu_and_ml/yolo/yolo.py

```diff
@@ -0,0 +1,288 @@
+
+@app.function(
+    image=yolo_image,
+    timeout=1500,
+    volumes={"/root/data/": volume},
+    gpu='a100',
+    cpu=2,
+)
+def train(model_name = "doclaynet-base", data_path = '/root/data/doclay-dataset/data.yaml'):
+    img_size = 896
+    model = load_yolo_weights()
+    model.train(
+        data=data_path,
+        epochs=8,
+        batch=0.8, # automatic batch size to target 80% util
+        workers=4,
+        cache=True,
+        imgsz=img_size,
+        name=model_name,
+        device=0,
+        verbose=True,
+        val=True,
+        plots=True,
+        save_period=1,
+        project='/root/data/runs',
+        fraction=0.4,
+        exist_ok=True, # Replace existing run
+    )
+
+# ## Running the Training Job
+# We call our Modal train function to start our training job.
+# The --detach flag is used so that the job continues in the cloud even if our terminal disconnects.
+# ```bash
+# modal run --detach yolo.py::train
+# ```
+# Once successful, Ultralytics will save the trained model in /root/data/runs/doclaynet-base/weights/best.pt
+# 
+# 
+# ### Optional: Running Additional Finetunes
+# If you'd like to train new models on other datasets you've downloaded into your Volume, 
+# you can reuse the train function as such:
+# ```bash
+# modal run --detach yolo.py::train --model_name=finetune_1 --data_path=/root/data/finetune1-dataset/data.yaml
+# ```
+
+# ## Creating Our Inference Class
+# 
+# We define a class-based remote handler for inference, allowing us to load our model only once 
+# on container boot and reuse it for future inferences.
+
+@app.cls(
+    image=yolo_image,
+    volumes={"/root/data/": volume},
+    gpu='h100',
+)
+class YOLOInference:
+    def __init__(self, weights_path):
+        self.weights_path = weights_path
+
+    @modal.enter()
+    def load_model(self):
+        # load model only once, on container boot
+        from ultralytics import YOLO
+        self.model = YOLO(self.weights_path)
+
+    @modal.method()
+    def predict(self, input_dir: str, output_dir: str):
+        import os
+        from tqdm import tqdm
+
+        # We'll be saving results into 
+        os.makedirs(output_dir, exist_ok=True)
+
+        # Get all image files from the input directory
+        image_files = [f for f in os.listdir(input_dir) if f.lower().endswith(('.png', '.jpg', '.jpeg'))]
+
+        for image_file in tqdm(image_files):
+            input_path = os.path.join(input_dir, image_file)
+            
+            # Perform inference
+            results = self.model.predict(input_path)
+            
+            # Save the results
+            output_path = os.path.join(output_dir, os.path.basename(image_file))
+            print("SAVING: ", output_path)
+            results[0].save(output_path)
+
+        print(f"Inference complete. Results saved to {output_dir}")
+    
+@app.local_entrypoint()
+def infer(model_name = "doclaynet-base", input_dir = "/root/data/doclay-dataset/images/val"):
```

Comment:
I generally prefer the `local_entrypoint` to run on its own, without requiring any other functions like `download_dataset` or `train` to run.

Reply:
And we call the `local_entrypoint` by `main` usually, because that's the purpose it serves.